crates/dbsp/src/trace/layers/column_layer/mod.rs

```diff
@@ -305,6 +306,30 @@ impl<K, R> ColumnLayer<K, R> {
         // All item are processed. This can be optimized to `set_len` by LLVM.
         drop(shifter);
     }
+
+    /// Compute a random sample of size `sample_size` of keys in `self.keys`.
+    ///
+    /// Pushes the random sample of keys to the `output` vector in ascending
+    /// order.
+    pub fn sample_keys<RG>(&self, rng: &mut RG, sample_size: usize, output: &mut Vec<K>)
+    where
+        K: Clone,
+        RG: Rng,
+    {
+        let size = self.len() - self.lower_bound;
+
+        if sample_size >= size {
+            for key in &self.keys[self.lower_bound..] {
+                output.push(key.clone());
+            }
+        } else {
+            let mut indexes = sample(rng, size, sample_size).into_vec();
+            indexes.sort_unstable();
+            for index in indexes.into_iter() {
+                output.push(self.keys[index + self.lower_bound].clone());
+            }
+        }
```

Comment:
Pre-allocate the output vector
```suggestion
        output.reserve(min(sample_size, size));

        if sample_size >= size {
            for key in &self.keys[self.lower_bound..] {
                output.push(key.clone());
            }
        } else {
            let mut indexes = sample(rng, size, sample_size).into_vec();
            indexes.sort_unstable();

            for index in indexes {
                output.push(self.keys[index + self.lower_bound].clone());
            }
        }
```